fwidgen/src/main.rs

```diff
@@ -0,0 +1,132 @@
+use anyhow::{anyhow, Context, Result};
+use clap::Parser;
+use hubtools::RawHubrisArchive;
+use sha3::{Digest, Sha3_256};
+use std::{
+    str,
+    ops::Range,
+};
+
+pub const LPC55_FLASH_PAGE_SIZE: usize = 512;
+
+#[derive(Parser, Debug)]
+struct Args {
+    /// Hubris archive
+    #[clap(env = "HUBEDIT_ARCHIVE")]
+    archive: String,
+}
+
+#[derive(Debug)]
+enum Chip {
+    Lpc55,
+    Stm32,
+}
+
+impl TryFrom<&RawHubrisArchive> for Chip {
+    type Error = anyhow::Error;
+
+    fn try_from(archive: &RawHubrisArchive) -> Result<Self> {
+        let manifest = archive.extract_file("app.toml")?;
+        let manifest: toml::Value = toml::from_str(
+            str::from_utf8(&manifest).context("manifest bytes to UTF8")?,
+        ).context("manifest UTF8 to TOML")?;
+
+        let chip = manifest
+           .as_table()
+           .ok_or(anyhow!("manifest isn't a table"))?
+           .get("chip")
+           .ok_or(anyhow!("no key \"chip\" in manifest"))?
+           .as_str()
+           .ok_or(anyhow!("value for key \"chip\" isn't a string"))?;
+
+        if chip.contains("lpc55") {
+            Ok(Chip::Lpc55)
+        } else if chip.contains("stm32") {
+            Ok(Chip::Stm32)
+        } else {
+            Err(anyhow!("Unsupported chip: {}", chip))
+        }
+    }
+}
+
+fn get_flash_range(name: &str, archive: &RawHubrisArchive) -> Result<Range<u32>> {
+    let memory = archive.extract_file("memory.toml")
+        .context("extract memory.toml from archive")?;
+    let memory: toml::Value = toml::from_str(
+        str::from_utf8(&memory).context("memory.toml bytes to UTF8")?,
+    ).context("memory.toml UTF8 to TOML")?;
+
+    // got carried away, there's probably a derive macro for TOML -> Rust
+    // types / instances
+    for value in memory
+        .as_table()
+        .ok_or(anyhow!("no table in memory.toml"))?
+        .get("flash")
+        .ok_or(anyhow!("no value with key \"flash\" memory.toml table"))?
+        .as_array()
+        .ok_or(anyhow!("value from key \"flash\" in memory.toml is not an array"))?
+    {
+        let name_toml = value
+            .get("name")
+            .ok_or(anyhow!("no key \"name\" found in memory.toml \"flash\""))?
+            .as_str()
+            .ok_or(anyhow!("value for key \"name\" isn't a string"))?;
+
+        if name == name_toml {
```

Comment:
It took me a minute to figure out/remember what this was doing, mostly because I thought there was only one flash range for the SP. Mind adding a comment explaining that we double check against the expected image name even though there is a single image?

Reply:
This is only one flash range for the SP but I didn't want to hard code assumptions especially since the generic approach isn't too much code. Either way I'm happy to add a comment to make this more clear.

Reply:
I've gone back through this and added a lot of context to the comments + a doc comment for `--help` output.